designer/client/src/components/graph/node-modal/editors/expression/Table/tableState.ts

```diff
@@ -0,0 +1,224 @@
+                    state.rows,
+                ),
+            };
+        case ActionTypes.renameColumn:
+            // prevent duplicates
+            if (state.columns.find(({ name }) => name === action.to)) {
+                return state;
+            }
+            return {
+                ...state,
+                columns: state.columns.map(({ name, ...col }) => ({ name: name === action.from ? action.to : name, ...col })),
+            };
+        case ActionTypes.deleteColumns:
+            return {
+                ...state,
+                columns: state.columns.filter((_, i) => !action.columns.includes(i)),
+                rows: state.rows.map((r) => r.filter((_, i) => !action.columns.includes(i))),
+            };
+        case ActionTypes.deleteRows:
+            return {
+                ...state,
+                rows: state.rows.filter((_, i) => !action.rows.includes(i)),
+            };
+        case ActionTypes.changeColumnType:
+            if (!(action.column >= 0 && action.dataType)) {
+                return state;
+            }
+            return {
+                ...state,
+                columns: state.columns.map((column, i) => {
+                    if (i !== action.column) {
+                        return column;
+                    }
+                    return { ...column, type: action.dataType };
+                }),
+            };
+        case ActionTypes.columnResize:
+            return {
+                ...state,
+                columns: state.columns.map?.((column, i) => {
+                    if (i !== action.column) {
+                        return column;
+                    }
+                    return { ...column, size: action.size.toString() };
+                }),
+            };
+        case ActionTypes.resetColumnsSize:
+            return {
+                ...state,
+                columns: state.columns.map(({ size, ...c }, i) => (action.columns.includes(i) ? c : { ...c, size })),
+            };
+        case ActionTypes.expand:
+            return expandTable(state, action.rows, action.columns, action.dataType);
+        case ActionTypes.replaceData:
+            return action.data;
+        default:
+            return state;
+    }
+}
+
+export function useTableState({ expression, language }: ExpressionObj): [TableData, Dispatch<Action>, string] {
+    const [rawExpression, setRawExpression] = useState<string>(expression);
+
+    const fromExpression = useMemo(() => getParser(language), [language]);
+    const toExpression = useMemo(() => getStringifier(language), [language]);
+
+    const [state, dispatch] = useReducer(reducer, emptyValue, (defaultValue) => fromExpression(expression, defaultValue));
+
+    useEffect(() => {
+        setRawExpression((current) => {
+            if (current === expression) {
+                return current;
+            }
+            dispatch({
```

Comment:
This whole logic looks magical to me, i needed a few seconds to understand what's going on and still don't know when this dispatcher will potentially call.